File: app/src/main/java/com/mappls/sdk/demo/java/plugin/DirectionPolylinePlugin.java
package com.mappls.sdk.demo.java.plugin;

import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineCap;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineColor;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineDasharray;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineJoin;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineWidth;

import android.graphics.Color;

import androidx.annotation.NonNull;

import com.mappls.sdk.geojson.Feature;
import com.mappls.sdk.geojson.FeatureCollection;
import com.mappls.sdk.geojson.LineString;
import com.mappls.sdk.geojson.Point;
import com.mappls.sdk.maps.MapView;
import com.mappls.sdk.maps.MapplsMap;
import com.mappls.sdk.maps.Style;
import com.mappls.sdk.maps.geometry.LatLng;
import com.mappls.sdk.maps.style.layers.Layer;
import com.mappls.sdk.maps.style.layers.LineLayer;
import com.mappls.sdk.maps.style.layers.Property;
import com.mappls.sdk.maps.style.sources.GeoJsonOptions;
import com.mappls.sdk.maps.style.sources.GeoJsonSource;
import com.mappls.sdk.maps.utils.ColorUtils;
import com.mappls.sdk.services.api.directions.DirectionsCriteria;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Saksham on 19/9/19.
 */
public class DirectionPolylinePlugin implements MapView.OnDidFinishLoadingStyleListener {
    private static final String UPPER_SOURCE_ID = "line-source-upper-id";

    private MapplsMap mapplsMap;

    private FeatureCollection featureCollection;
    private static final String LAYER_ID = "line-layer-upper-id";
    private List<LatLng> latLngs;

    private float widthDash = 5f;
    private float gapDash = 5f;

    private String directionsCriteria;

    private GeoJsonSource polylineSource;
    private LineLayer lineLayer;

    public DirectionPolylinePlugin(MapplsMap mapplsMap, MapView mapView, String directionsCriteria) {
        this.mapplsMap = mapplsMap;
        this.directionsCriteria = directionsCriteria;

        updateSource();
        mapView.addOnDidFinishLoadingStyleListener(this);
    }

    /**
     * Add polyline features and set polyline property for walk and other
     *
     * @param latLngs list of points
     */
    public void createPolyline(List<LatLng> latLngs) {
        this.latLngs = latLngs;
        mapplsMap.getStyle(new Style.OnStyleLoaded() {
            @Override
            public void onStyleLoaded(@NonNull Style style) {
                List<Point> points = new ArrayList<>();
                for (LatLng latLng : latLngs) {
                    points.add(Point.fromLngLat(latLng.getLongitude(), latLng.getLatitude()));
                }
                Feature features = Feature.fromGeometry(LineString.fromLngLats(points));
                Layer layer = style.getLayer(LAYER_ID);
                if (layer != null) {
                    if (directionsCriteria.equalsIgnoreCase(DirectionsCriteria.PROFILE_WALKING)) {

                        layer.setProperties(lineDasharray(new Float[]{widthDash, gapDash}),
                                lineColor(Color.BLACK));
                    } else {
                        layer.setProperties(lineDasharray(new Float[]{}),
                                lineColor(ColorUtils.colorToRgbaString(Color.parseColor("#3bb2d0"))));
                    }
                }
                featureCollection = FeatureCollection.fromFeature(features);
                initSources(featureCollection);
            }
        });

    }

    /**
     * Update polyline features and set polyline property for walk and other
     *
     * @param directionsCriteria {"foot", "biking", "driving"}
     * @param latLngs            list of points
     */
    public void updatePolyline(String directionsCriteria, List<LatLng> latLngs) {
        this.directionsCriteria = directionsCriteria;
        List<Point> points = new ArrayList<>();
        for (LatLng latLng : latLngs) {
            points.add(Point.fromLngLat(latLng.getLongitude(), latLng.getLatitude()));
        }
        Feature features = Feature.fromGeometry(LineString.fromLngLats(points));
        if (directionsCriteria.equalsIgnoreCase(DirectionsCriteria.PROFILE_WALKING)) {
            lineLayer.setProperties(lineDasharray(new Float[]{widthDash, gapDash}),
                    lineColor(Color.BLACK));
        } else {
            lineLayer.setProperties(lineDasharray(new Float[]{}),
                    lineColor(ColorUtils.colorToRgbaString(Color.parseColor("#3bb2d0"))));
        }

        featureCollection = FeatureCollection.fromFeature(features);
        updateSource();
    }

    /**
     * Add various sources to the map.
     */
    private void initSources(@NonNull FeatureCollection featureCollection) {
        mapplsMap.getStyle(new Style.OnStyleLoaded() {
            @Override
            public void onStyleLoaded(@NonNull Style style) {
                if (style.getSource(UPPER_SOURCE_ID) == null) {
                    style.addSource(polylineSource = new GeoJsonSource(UPPER_SOURCE_ID, featureCollection,
                            new GeoJsonOptions().withLineMetrics(true).withBuffer(2)));
                }
            }
        });

    }

    /**
     * Update Source and GeoJson properties
     */
    private void updateSource() {
        mapplsMap.getStyle(new Style.OnStyleLoaded() {
            @Override
            public void onStyleLoaded(@NonNull Style style) {
                GeoJsonSource source = (GeoJsonSource) style.getSource(UPPER_SOURCE_ID);
                if (source == null) {
                    create(style);
                    return;
                }
                if (featureCollection != null) {
                    polylineSource.setGeoJson(featureCollection);
                }
            }
        });

    }

    /**
     * Add Line layer on map
     */
    private void create(Style style) {
        if (style.getLayer(LAYER_ID) == null) {
            style.addLayer(lineLayer = new LineLayer(LAYER_ID, UPPER_SOURCE_ID).withProperties(
                    lineCap(Property.LINE_CAP_ROUND),
                    lineJoin(Property.LINE_JOIN_ROUND),
                    lineWidth(5f)));


            if (directionsCriteria.equalsIgnoreCase(DirectionsCriteria.PROFILE_WALKING)) {
                lineLayer.setProperties(lineDasharray(new Float[]{gapDash, widthDash}));
            }
        }
    }


    @Override
    public void onDidFinishLoadingStyle() {
        updateSource();
        createPolyline(latLngs);
    }
}

File: app/src/main/java/com/mappls/sdk/demo/kotlin/plugin/DirectionPolylinePlugin.kt
package com.mappls.sdk.demo.kotlin.plugin

import android.graphics.Color
import com.mappls.sdk.geojson.Feature

import com.mappls.sdk.geojson.FeatureCollection
import com.mappls.sdk.geojson.LineString
import com.mappls.sdk.geojson.Point
import com.mappls.sdk.maps.MapView
import com.mappls.sdk.maps.MapplsMap
import com.mappls.sdk.maps.Style
import com.mappls.sdk.maps.geometry.LatLng
import com.mappls.sdk.maps.style.layers.LineLayer
import com.mappls.sdk.maps.style.layers.Property
import com.mappls.sdk.maps.style.layers.PropertyFactory.*
import com.mappls.sdk.maps.style.sources.GeoJsonOptions
import com.mappls.sdk.maps.style.sources.GeoJsonSource
import com.mappls.sdk.maps.utils.ColorUtils
import com.mappls.sdk.services.api.directions.DirectionsCriteria
import java.util.*

/**
 * Created by Saksham on 20/9/19.
 */
class DirectionPolylinePlugin(val mapplsMap: MapplsMap, mapView: MapView, private var directionsCriteria: String?) : MapView.OnDidFinishLoadingStyleListener {

    private var featureCollection: FeatureCollection? = null
    private var latLngs: List<LatLng>? = null

    private val widthDash = 5f
    private val gapDash = 5f

    private var polylineSource: GeoJsonSource? = null
    private var lineLayer: LineLayer? = null

    init {

        updateSource()
        mapView.addOnDidFinishLoadingStyleListener(this)
    }

    /**
     * Add polyline features and set polyline property for walk and other
     *
     * @param latLngs list of points
     */
    fun createPolyline(latLngs: List<LatLng>) {
        this.latLngs = latLngs
        val points = ArrayList<Point>()
        for (latLng in latLngs) {
            points.add(Point.fromLngLat(latLng.longitude, latLng.latitude))
        }
        val features = Feature.fromGeometry(LineString.fromLngLats(points))

        if (directionsCriteria!!.equals(DirectionsCriteria.PROFILE_WALKING, ignoreCase = true)) {
            lineLayer!!.setProperties(lineDasharray(arrayOf(widthDash, gapDash)),
                    lineColor(Color.BLACK))
        } else {
            lineLayer!!.setProperties(lineDasharray(arrayOf()),
                    lineColor(ColorUtils.colorToRgbaString(Color.parseColor("#3bb2d0"))))
        }
        featureCollection = FeatureCollection.fromFeature(features)
        initSources(featureCollection!!)
    }

    /**
     * Update polyline features and set polyline property for walk and other
     *
     * @param directionsCriteria {"foot", "biking", "driving"}
     * @param latLngs list of points
     */
    fun updatePolyline(directionsCriteria: String, latLngs: List<LatLng>) {
        this.directionsCriteria = directionsCriteria
        val points = ArrayList<Point>()
        for (latLng in latLngs) {
            points.add(Point.fromLngLat(latLng.longitude, latLng.latitude))
        }
        val features = Feature.fromGeometry(LineString.fromLngLats(points))
        if (directionsCriteria.equals(DirectionsCriteria.PROFILE_WALKING, ignoreCase = true)) {
            lineLayer!!.setProperties(lineDasharray(arrayOf(widthDash, gapDash)),
                    lineColor(Color.BLACK))
        } else {
            lineLayer!!.setProperties(lineDasharray(arrayOf()),
                    lineColor(ColorUtils.colorToRgbaString(Color.parseColor("#3bb2d0"))))
        }

        featureCollection = FeatureCollection.fromFeature(features)
        updateSource()
    }

    /**
     * Add various sources to the map.
     */
    private fun initSources(featureCollection: FeatureCollection) {
        mapplsMap.getStyle {
            if(it.getSource(UPPER_SOURCE_ID) == null) {
                polylineSource = GeoJsonSource(UPPER_SOURCE_ID, featureCollection,
                        GeoJsonOptions().withLineMetrics(true).withBuffer(2))
                it.addSource(polylineSource!!)
            }
        }

    }

    /**
     * Update Source and GeoJson properties
     */
    private fun updateSource() {
        mapplsMap.getStyle {
            val source = it.getSource(UPPER_SOURCE_ID) as GeoJsonSource?
            if (source == null) {
                create(it)
                return@getStyle
            }
            if (featureCollection != null) {
                source.setGeoJson(featureCollection)
            }
        }

    }

    /**
     * Add Line layer on map
     */
    private fun create(style: Style) {
        if(style.getLayer(LAYER_ID) == null) {
            lineLayer = LineLayer(LAYER_ID, UPPER_SOURCE_ID).withProperties(
                    lineCap(Property.LINE_CAP_ROUND),
                    lineJoin(Property.LINE_JOIN_ROUND),
                    lineWidth(5f))
            style.addLayer(lineLayer!!)


            if (directionsCriteria!!.equals(DirectionsCriteria.PROFILE_WALKING, ignoreCase = true)) {
                lineLayer?.setProperties(lineDasharray(arrayOf(gapDash, widthDash)))
            }
        }
    }


    override fun onDidFinishLoadingStyle() {
        updateSource()
        createPolyline(latLngs!!)
    }

    companion object {
        private const val UPPER_SOURCE_ID = "line-source-upper-id"
        private const val LAYER_ID = "line-layer-upper-id"
    }
}